### PR TITLE
feat: support Copilot CLI frontmatter fields (user-invocable, disable-model-invocation)

### PR DESCRIPTION
**Pull request by charris-msft**
Add support for new Copilot CLI skill frontmatter fields:
- user-invocable (boolean): whether users can invoke via /skill-name
- disable-model-invocation (boolean): prevent model from invoking skill
- allowed-tools: validate format when present

New checks in score.ts:
- spec-user-invocable: validates boolean value when present
- spec-disable-model-invocation: validates boolean value when present
- spec-allowed-tools: validates non-empty when present
- reference-only-pattern: detects user-invocable=false + disable-model-invocation=true (creates a reference file loaded from disk but never invoked)

Updated ALLOWED_FIELDS to prevent false positive warnings on valid fields. Updated scoring.md and SKILL.md documentation with new field patterns.

Co-authored-by: Copilot <223556219+Copilot@users.noreply.github.com>

### CODE REVIEW

**Comment by spboyer** [comment]
@charris-msft  These are not in the official skills spec. How would a user know when or what these options or guidance to use or not. Can you supply reference link for these fields spec?

**Comment by charris-msft** [comment]
Good call — these fields are **not** in the agentskills.io spec; they're Copilot CLI extensions documented here:

📎 [Custom agents configuration — YAML frontmatter properties](https://docs.github.com/copilot/reference/custom-agents-configuration#yaml-frontmatter-properties)

I've pushed a commit (9af511d) that adds this reference link in two places:
- **references/scoring.md** — the 'Copilot CLI Extension Fields' section now links to the GitHub docs and clarifies these are not part of the agentskills.io spec
- **scripts/src/tokens/commands/score.ts** — added the docs URL alongside the existing spec reference in the header comment

All 109 tests pass and token limits are within budget. ✅

**Comment by charris-msft** [comment]
> @charris-msft These are not in the official skills spec. How would a user know when or what these options or guidance to use or not. Can you supply reference link for these fields spec?

This looks like the right location:
https://docs.github.com/copilot/reference/custom-agents-configuration#yaml-frontmatter-properties

It only has one of the properties right now, but the other should show up here once the GH team adds it.

**Review by spboyer — COMMENTED** [review]
Two items from review:

### 1. Bug: ALLOWED_FIELDS may not include the new fields

The PR description says "Updated ALLOWED_FIELDS to prevent false positive warnings" -- but please verify `user-invocable` and `disable-model-invocation` are actually in the ALLOWED_FIELDS Set. If they're missing, any skill using these valid Copilot CLI fields will trigger a false `spec-allowed-fields` warning ("Unknown frontmatter fields"), even though the PR adds dedicated check functions for them.

### 2. Scoring posture: these are Copilot-specific, not global spec

`user-invocable`, `disable-model-invocation`, and `allowed-tools` are **Copilot CLI extensions** -- not part of the agentskills.io specification. Their absence should **not** negatively affect scoring. The checks should be purely informational:

- Present and valid -> note it (optimal or ok)
- Not present -> silent pass (not a warning)
- Present but invalid value -> warn (e.g. non-boolean for user-invocable)

The current implementation looks like it returns `ok` when absent, which is correct -- but the `spec-` prefix on the check names implies these are spec requirements. Consider either:
- Renaming to `copilot-user-invocable` / `copilot-disable-model-invocation` / `copilot-allowed-tools` to make it clear these are platform extensions
- Or adding a note in the output/docs that these checks enhance Copilot experience but are not required for spec compliance

Nice work overall -- the reference-only pattern detection is a useful addition.

**Comment by charris-msft** [comment]
**Correction**: The link I added earlier was wrong — it pointed to the custom agents configuration docs (.agent.md), not skills.

After checking the [Copilot agent runtime source](https://github.com/github/copilot-agent-runtime/blob/main/src/skills/types.ts), these fields **are valid for SKILL.md** — they're defined in the `skillFrontmatterSchema`:

| Field | SKILL.md | Commands (.md) | Custom Agents (.yaml) |
|:------|:--------:|:--------------:|:---------------------:|
| `user-invocable` | ✅ Skills only | ❌ | ❌ |
| `disable-model-invocation` | ✅ | ✅ | ❌ |
| `allowed-tools` | ✅ | ✅ | ❌ |

`scoring.md`: These fields are part of the Copilot skill frontmatter schema (not yet documented publicly; not part of the agentskills.io spec)

`score.ts`: Copilot CLI fields: not yet documented publicly (see skillFrontmatterSchema in copilot-agent-runtime)

**Comment by charris-msft** [comment]
Addressed both review items:

### 1. ALLOWED_FIELDS
Confirmed -- `user-invocable` and `disable-model-invocation` were already in the ALLOWED_FIELDS Set. No false positives.

### 2. Renamed spec- to copilot- prefix
Renamed all Copilot-specific checks to make the distinction self-documenting:

| Before | After |
|--------|-------|
| spec-user-invocable | copilot-user-invocable |
| spec-disable-model-invocation | copilot-disable-model-invocation |
| spec-allowed-tools | copilot-allowed-tools |
| reference-only-pattern | copilot-reference-only-pattern |

Updated in score.ts, score.test.ts, and scoring.md. All 109 tests pass. (cb2c53f)

**Comment by spboyer** [comment]
@charris-msft some merge conflicts to check

**Comment by charris-msft** [comment]
@spboyer - conflicts resolved. Thanks for the heads up.

File: scripts/src/tokens/commands/score.ts [hunk 1]
```diff
@@ -334,6 +342,129 @@ export function checkVersionRecommendation(fields: Record<string, unknown>): Adv
...
+  if (!trimmed) {
+    return {
+      name: 'copilot-allowed-tools',
+      status: 'warning',
+      message: 'allowed-tools is empty — omit the field or specify tool names',
+      evidence: 'Copilot CLI skill frontmatter: comma-separated list of tool names'
+    };
+  }
+
+  return {
```
Comment (spboyer) [inline scripts/src/tokens/commands/score.ts:444]: **[MEDIUM]** Logic: `checkAllowedToolsFormat` only rejects empty values, so malformed values like `shell read write` or `shell,,read` still pass even though this PR documents the field as a comma-separated list. That makes the new Copilot field check report success for invalid frontmatter. Split on commas and warn when entries are empty, or when a multi-tool value has no comma separators.

**Review by spboyer — CHANGES_REQUESTED** [review]
This is close, and the prior thread items are mostly addressed: the new fields are in the allowlist, the Copilot checks were renamed away from spec-* names, CI is green, and the merge conflicts are resolved.

Issues to address:
- scripts/src/tokens/commands/score.ts:406 - allowed-tools validation accepts any non-empty string despite documenting comma-separated validation

**Comment by charris-msft** [comment]
Addressed the remaining review item. llowed-tools now validates comma-separated entries instead of accepting any non-empty string: it strips surrounding scalar quotes, rejects empty entries from leading/trailing/repeated commas, and warns on whitespace-separated tool names like Bash Read. I also merged upstream main; the PR now reports mergeStateStatus=CLEAN.

**Review by jongio — COMMENTED** [review]
The main issue from @spboyer's prior review (allowed-tools accepting invalid formats) was addressed in commit 2ae95f8 - `checkAllowedToolsFormat` now correctly warns on space-separated entries (`shell read write`) and empty entries from repeated/trailing commas (`shell,,read`).

Remaining items:

1. **Whitespace diff noise** - Large chunks of score.ts show trailing-whitespace normalization (CRLF/LF). Not wrong, but inflates the diff and makes future blame harder to read. Consider squashing these into a single `chore: normalize line endings` commit or stripping them from this PR.

2. **Scoring pseudocode rewrite (scoring.md)** - The algorithm documentation was simplified from explicit helper functions to a compact version. The logic is equivalent, but the original was more self-documenting for contributors who need to understand the scoring tiers. If this was intentional simplification, it's fine - just noting the documentation lost some clarity.

3. **The Copilot bot's js-yaml concern doesn't apply here** - The `parseFrontmatter` function is a custom regex parser that always produces string values. Unquoted YAML booleans won't be silently converted to `boolean` type since `js-yaml` isn't used for frontmatter parsing.

The feature is well-structured, tests cover the important edge cases, and the `copilot-` prefix naming (per @spboyer's earlier feedback) correctly separates these from spec-mandated checks.

File: scripts/src/tokens/commands/score.ts (hunk 1 above)
Comment (spboyer) [reply scripts/src/tokens/commands/score.ts:444]: Fixed in 8486400 — @jongio confirmed the whitespace detection was added in commit 2ae95f8. The fix is already in the branch.

File: scripts/src/tokens/commands/score.ts
[the remark this answers is not in the document]
Comment (spboyer) [reply scripts/src/tokens/commands/score.ts:67]: Not applicable — this parser uses custom regex (not js-yaml) and always produces string values. @jongio confirmed this. New tests in commit 8486400 prove unquoted booleans (true/false/yes/no) are preserved as strings.

File: scripts/src/tokens/commands/score.ts [hunk 2]
```diff
@@ -95,27 +100,30 @@ export function parseFrontmatter(content: string): ParsedFrontmatter | null {
...
-      currentValue = isBlockScalarMarker(rawVal) ? '' : rawVal;
+    const keyMatch = line.match(/^(\w[\w-]*)\s*:\s*(.*)/);
+    if (keyMatch) {
+      if (currentKey) {
+        fields[currentKey] = normalizeFrontmatterValue(currentValue);
+      }
+      currentKey = keyMatch[1];
+      const rawVal = keyMatch[2].trim();
+      currentValue = isBlockScalarMarker(rawVal) ? '' : rawVal;
     } else if (currentKey && (line.startsWith('  ') || line.startsWith('\t'))) {
```
[the remark this answers is not in the document]
Comment (spboyer) [reply scripts/src/tokens/commands/score.ts:111]: Not applicable — this parser uses custom regex (not js-yaml) and always produces string values. @jongio confirmed this. New tests in commit 8486400 prove unquoted booleans are preserved as strings.

File: scripts/src/tokens/commands/score.ts [hunk 3]
```diff
@@ -18,13 +19,14 @@ const MAX_DESCRIPTION_LENGTH = 1024;
 /** Max compatibility field length per agentskills.io spec */
 const MAX_COMPATIBILITY_LENGTH = 500;
 
-/** Allowed frontmatter fields per agentskills.io spec */
+/** Allowed frontmatter fields per agentskills.io spec + Copilot CLI extensions */
 const ALLOWED_FIELDS = new Set([
-  'name', 'description', 'license', 'allowed-tools', 'metadata', 'compatibility'
+  'name', 'description', 'license', 'allowed-tools', 'metadata', 'compatibility',
+  'user-invocable', 'disable-model-invocation'
 ]);
```
[the remark this answers is not in the document]
Comment (spboyer) [reply scripts/src/tokens/commands/score.ts:26]: Fixed in 8486400 — updated evidence string and success message to clarify the field list includes Copilot CLI extensions.

File: scripts/src/tokens/commands/score.ts
[the remark this answers is not in the document]
Comment (spboyer) [reply scripts/src/tokens/commands/score.ts:348]: Fixed in 8486400 — warning message now explicitly lists all accepted values: true/false/yes/no/on/off.

File: scripts/src/tokens/commands/score.ts [hunk 4]
```diff
@@ -350,30 +358,175 @@ export function checkVersionRecommendation(fields: Record<string, unknown>): Adv
...
+  'true', 'false', 'yes', 'no', 'on', 'off'
+]);
+
+/**
+ * Copilot CLI Check: Validate that a frontmatter field is a boolean value when present.
+ */
+export function checkBooleanField(fieldName: string, value: string | undefined): AdvisoryCheck {
+  const checkName = `copilot-${fieldName}`;
+
+  if (value === undefined) {
```
[the remark this answers is not in the document]
Comment (spboyer) [reply scripts/src/tokens/commands/score.ts:372]: Fixed in 8486400 — warning message now explicitly lists all accepted values: true/false/yes/no/on/off.

File: scripts/src/tokens/commands/score.test.ts [hunk 5]
```diff
@@ -385,11 +388,179 @@ describe('checkAllowedFields', () => {
...
+    const fields = { name: 'x', description: 'y', 'disable-model-invocation': 'true' };
+    const result = checkAllowedFields(fields);
+    expect(result.status).toBe('ok');
+  });
+
+  it('allows all Copilot CLI fields together', () => {
+    const fields = {
+      name: 'x', description: 'y', 'allowed-tools': 'Bash',
+      'user-invocable': 'true', 'disable-model-invocation': 'false'
+    };
```
[the remark this answers is not in the document]
Comment (spboyer) [reply scripts/src/tokens/commands/score.test.ts:430]: Fixed in 8486400 — added 4 new parseFrontmatter tests verifying unquoted booleans (true, false, yes, no) are preserved as strings.

File: scripts/src/tokens/commands/score.test.ts [hunk 6]
```diff
@@ -385,11 +388,179 @@ describe('checkAllowedFields', () => {
...
+    const result = checkBooleanField('user-invocable', 'no');
+    expect(result.status).toBe('ok');
+  });
+
+  it('warns on non-boolean value', () => {
+    const result = checkBooleanField('user-invocable', 'maybe');
+    expect(result.status).toBe('warning');
+    expect(result.message).toContain('boolean');
+    expect(result.message).toContain('maybe');
+  });
```
[the remark this answers is not in the document]
Comment (spboyer) [reply scripts/src/tokens/commands/score.test.ts:474]: Fixed in 8486400 — added 4 new parseFrontmatter tests verifying unquoted booleans (true, false, yes, no) are preserved as strings.

File: SKILL.md [hunk 7]
```diff
@@ -168,6 +168,9 @@ Template (with routing clarity for High score):
 ---
 name: skill-name
 description: "**WORKFLOW SKILL** — [ACTION VERB] [UNIQUE_DOMAIN]. [Clarifying sentence]. WHEN: \"[phrase1]\", \"[phrase2]\", \"[phrase3]\". INVOKES: [tools/MCP servers used]. FOR SINGLE OPERATIONS: [when to bypass this skill]."
+allowed-tools: shell, read, write    # optional: auto-allowed tools
```
Comment (spboyer) [inline SKILL.md:171]: **[MEDIUM]** Documentation: The template embeds inline YAML comments on the new Copilot fields, but `parseFrontmatter` keeps trailing `# ...` text in the scalar value. Users copying this example get values like `true # optional...` and `write # optional...`, which makes `checkBooleanField` and `checkAllowedToolsFormat` warn on the template itself. Move the comments into prose outside the YAML block, or teach the parser to strip unquoted trailing YAML comments and add a regression test.

File: scripts/src/tokens/commands/score.ts [hunk 8]
```diff
@@ -350,30 +358,175 @@ export function checkVersionRecommendation(fields: Record<string, unknown>): Adv
...
 
+/** Valid boolean string values in YAML frontmatter */
+const YAML_BOOLEAN_VALUES = new Set([
+  'true', 'false', 'yes', 'no', 'on', 'off'
+]);
+
+/**
+ * Copilot CLI Check: Validate that a frontmatter field is a boolean value when present.
+ */
+export function checkBooleanField(fieldName: string, value: string | undefined): AdvisoryCheck {
```
Comment (spboyer) [inline scripts/src/tokens/commands/score.ts:369]: **[LOW]** Public API conventions: These new Copilot CLI check helpers are exported from `score.ts`, but they are not available from the package's `@spboyer/sensei/checks` entry point like the existing individual check helpers. Consumers can run the aggregate scorer but cannot import the new checks from the checks API. Add `checkBooleanField`, `checkAllowedToolsFormat`, and `checkReferenceOnlyPattern` to `scripts/src/checks.ts` and cover that export.

**Review by spboyer — CHANGES_REQUESTED** [review]
The previous allowed-tools validation issue is addressed, and the changed scoring reference is back under its token limit. One new template/docs issue remains because the SKILL.md example now produces values that fail the validators it documents.

Issues to address:
- SKILL.md:171 - inline YAML comments in the template are parsed as scalar content and make the example fail validation
- scripts/src/tokens/commands/score.ts:369 - new Copilot helper checks are not exposed through the checks entry point
